### PR TITLE
[Model] Add F5-TTS support

### LINKED ISSUE #1551: [New Model] F5-TTS (flow-matching DiT-based TTS)
## Model Information
- **Model**: F5-TTS v1
- **Type**: TTS (Text-to-Speech)
- **HuggingFace**: https://huggingface.co/SWivid/F5-TTS
- **Paper/Repo**: https://github.com/SWivid/F5-TTS

## Why support this model?
F5-TTS is one of the most popular open-source TTS models with 14k+ GitHub stars. It uses a fully non-autoregressive flow-matching approach with a DiT backbone — architecturally similar to diffusion models already in vllm-omni. Its simplicity (no duration model, no phoneme alignment, no separate text encoder) and fast inference (RTF 0.15) make it well-suited for high-throughput serving.

## Technical Notes
- DiT architecture with flow matching (non-autoregressive)
- Text is padded with filler tokens to match speech length, then denoised
- ConvNeXt module refines text representation for alignment
- Sway Sampling inference strategy for improved quality/speed
- v1 base model released 2025/03/12 with improved training and inference
- Zero-shot voice cloning with short reference audio
- Standalone implementation (not yet in diffusers); PyPI package `f5-tts` available

## Closest supported model in vllm-omni
**Stable Audio Open** — also a diffusion-based audio generation model. F5-TTS's DiT + flow matching architecture is also conceptually similar to the image DiT models (FLUX, SD3) already supported, since the core denoising loop follows the same pattern but operates on mel spectrograms instead of pixel latents.

### PR DESCRIPTION
**Pull request by BeatSeat**
## Purpose

Adds native F5-TTS support: a flow-matching DiT TTS model (`SWivid/F5-TTS`
checkpoints) served through `/v1/audio/speech` in both offline and online
modes.

Closes #1551. Supersedes #2314.

## Summary

- **Model**: `F5TTSDiTModel` (22-layer DiT, AdaLayerNormZero modulation,
  shared attention backend, RoPE), vocos/BigVGAN vocoders, pinyin/char
  tokenizers. `F5TTSPipeline` owns mel front-end, flow-matching Euler
  sampling with sway sampling + CFG, and vocoder decode.
- **Serving**: `/v1/audio/speech` via the diffusion path
  (`guidance_scale`, `num_inference_steps`, `seed`, `ref_audio`/
  `ref_text` request fields); the F5 dispatch lives next to the existing
  diffusion branch until pure-diffusion TTS is supported by
  `tts_adapters`.
- **Loading**: bespoke discovery for the non-standard
  `SWivid/F5-TTS/<variant>` layout (synthetic transformer config,
  immediate vocab download, subfolder-prefixed weight patterns,
  `.pt`/safetensors routing).
- **Acceleration**: Cache-DiT (declarative), DiT CUDA Graph replay
  (opt-in via `--stage-overrides`; auto-gated under TP/HSDP and when a
  cache backend is active), FP8 plumbing. Measured on L4: Cache-DiT
  **1.62x RTF** (0.480 -> 0.294) with zero quality loss; TeaCache is
  fundamentally incompatible with flow-matching + sway sampling — do
  not use with F5-TTS.

## Test Plan

**vLLM Version:** 0.28.0

- Unit: `pytest tests/diffusion/models/f5_tts/test_hf_utils.py`
  (spec/vocab/config discovery — CPU-safe).
- Serving unit: `pytest
  tests/entrypoints/openai_api/test_serving_speech.py -k f5`
  (detection branch — CPU-safe).
- E2E (GPU): `pytest tests/e2e/online_serving/test_f5_tts.py`
  (L4, 3 cases: voice clone, custom params, Cache-DiT; needs CI run).
- E2E (GPU, manual): offline example
  `examples/offline_inference/text_to_speech/f5_tts/end2end.py`.

## Test Result

Measured on Colab L4, seed-tts-eval EN (200 samples/subset, WER/CER/SIM;
18 configs including an exhaustive TeaCache sweep):

| Config | RTF | Speedup | WER delta | Verdict |
|--------|----:|--------:|----------:|---------|
| Cache-DiT Fn1/Bn0/W4 (bf16) | 0.294 | **1.62x** | -4.5% (better) | **Recommended** |
| Baseline (bf16 autocast) | 0.480 | 1.00x | ref | Reference |
| TeaCache (any of 48+ configs) | — | 0.95-2.86x | +7..+416% | Incompatible |
| FP8 (any backend combo) | — | 0.86-0.91x | — | Skip on L4 |

- **TeaCache**: exhaustive sweep (48+ configs) proved step-level caching
  is fundamentally incompatible with flow-matching + sway sampling —
  quality-preserving configs are slower than baseline. No TeaCache
  change ships in this PR; do not use TeaCache with F5-TTS.
- **Cache-DiT**: block-level caching skips redundant blocks within a
  step (not whole steps) — 1.62x with zero quality loss.
- **FP8**: 10-17% slower on 0.3B+L4 (compute-bound). Kept as plumbing
  for H200-class validation later.
- **CUDA Graph**: implemented and hardened — opt-in via
  `--stage-overrides`, auto-gated under TP/HSDP and cache backends,
  side-stream warmup and thread-local capture errors per the MOSS
  pattern. L4 is compute-bound so the gain there is modest by
  construction; the launch-overhead share it targets grows on H200,
  where it is benchmarked next (Future work).

## Future work

- H200 validation: CUDA-graph x FP8 matrix; FP8 revisit — the L4
  regression is expected to flip on a memory-bound class part;
  CUDA-graph benchmark on the launch-overhead share that grows on
  faster GPUs.
- CI: the 3 e2e cases need a GPU CI run; register F5-TTS in
  `benchmarks/tts/model_configs.yaml` and
  `tests/dfx/perf/tests/test_tts.json` so the TTS benchmark and dfx
  perf suites cover it.
- Frame-wise streaming (chunked vocoder decode), per the ask in #2314.
- Serving: migrate the F5 dispatch from the `_is_f5_tts_model()`
  name-check to the `tts_adapters` framework once pure-diffusion
  models are supported there, and add PARITY-reference comments in
  `deploy/f5_tts.yaml` per the Nemotron convention (#6354).

## Cmd to run

```bash
vllm serve SWivid/F5-TTS/F5TTS_v1_Base --omni --trust-remote-code --port 8091

# Optional: enable the DiT CUDA Graph replay (auto-gated under TP/HSDP
# and when a cache backend is active; reference values in
# vllm_omni/deploy/f5_tts.yaml)
vllm serve SWivid/F5-TTS/F5TTS_v1_Base --omni --trust-remote-code \
    --stage-overrides '{"0":{"extras":{"f5_dit_cudagraph":true}}}'
```

(3-segment model id is required, e.g. `SWivid/F5-TTS/F5TTS_v1_Base`.
F5-TTS is not in the pipeline registry, so no deploy config is
auto-loaded; extras go through `--stage-overrides`.)

### CODE REVIEW

**Comment by zzehli** [comment]
@BeatSeat you should fix the ruff error

**Comment by BeatSeat** [comment]
@zzehli Fixed formatting and type annotations in commit 05d2d40bf (`audio_utils.py` `AudioSegment` type annotation and ruff format alignment across all touched files). All `ruff check` and `ruff format --check` inspections now pass cleanly.

File: vllm_omni/diffusion/models/f5_tts/pipeline_f5_tts.py [hunk 1]
```diff
@@ -0,0 +1,804 @@
...
+    primary: object | None,
+    fallback: object | None = None,
+    *,
+    default: object | None = None,
+) -> object | None:
+    """Normalize request fields that may arrive as a scalar or single-item list."""
+    value = primary if primary is not None else fallback
+    if value is None:
+        return default
+    if isinstance(value, (list, tuple)):
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/f5_tts/pipeline_f5_tts.py]: Keep decoded audio tuples intact; only unwrap single-item lists.

File: vllm_omni/diffusion/models/f5_tts/f5_tts_transformer.py [hunk 2]
```diff
@@ -0,0 +1,881 @@
...
+        value = value.view(batch_size, seq_len, self.num_kv_heads, self.head_dim)
+
+        # Apply rotary embeddings via shared RotaryEmbedding
+        if rope_emb is not None:
+            cos, sin = rope_emb
+            cos = cos.to(device=query.device, dtype=query.dtype)
+            sin = sin.to(device=query.device, dtype=query.dtype)
+            if self.pe_attn_head is not None:
+                # Partial RoPE: only first pe_attn_head heads
+                rotary_head_count = min(self.pe_attn_head, self.num_heads)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/f5_tts/f5_tts_transformer.py]: Apply partial RoPE using global head indices so TP ranks don't rotate extra heads.

File: vllm_omni/diffusion/models/f5_tts/audio_utils.py [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+
+    if audio.ndim == 1:
+        audio = audio.unsqueeze(0)  # [1, T]
+    else:
+        audio = audio.T  # [channels, T]
+
+    # RMS normalization
+    audio_rms = torch.sqrt(torch.mean(torch.square(audio))).item()
+    if target_rms is not None and audio_rms < target_rms:
+        audio = audio * target_rms / audio_rms
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/f5_tts/audio_utils.py:386]: Reject silent reference audio before dividing by its RMS.

File: vllm_omni/diffusion/models/f5_tts/pipeline_f5_tts.py [hunk 4]
```diff
@@ -0,0 +1,804 @@
...
+        cond_mel_len = cond_mel.shape[0]
+        total_mel_len = estimate_duration(cond_mel_len, cond_text_str, target_text, speed)
+        total_mel_len = max(total_mel_len, len(text_token_ids), cond_mel_len) + 1
+
+        assert len(text_token_ids) <= total_mel_len, (
+            f"number of text tokens ({len(text_token_ids)}) must be <= number of mel frames ({total_mel_len})"
+        )
+
+        # Padding and batching
+        cond_audio, cond_text_tensor, seq_len, _ = pad_and_batch(cond_mel, text_token_ids, total_mel_len)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/f5_tts/pipeline_f5_tts.py]: Reject requests exceeding 8192 mel frames before allocating sampling tensors.

**Comment by BeatSeat** [comment]
Thanks @lishunyang12 for the review!

We have pushed an update addressing all points:
1. **Audio tuple preservation**: Updated `_resolve_request_value` to only unwrap single-item lists, keeping decoded audio tuple structures intact.
2. **Partial RoPE TP fix**: Used global head indices (`tp_rank * num_heads + local_h < pe_attn_head`) so non-zero TP ranks do not rotate unintended local heads.
3. **Silent audio validation**: Added explicit check in `audio_utils.py` to reject silent reference audio with `audio_rms < 1e-6` before RMS normalization, preventing division by zero.
4. **Duration limit validation**: Added early rejection for requests exceeding 8192 estimated mel frames before allocating sampling tensors.
